samples/net/google_iot_mqtt/src/protocol.c

```diff
@@ -0,0 +1,370 @@
+/* Protocol implementation. */
+/*
+ * Copyright (c) 2018-2019 Linaro Ltd
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(net_google_iot_mqtt, LOG_LEVEL_DBG);
+#include "protocol.h"
+
+#include <zephyr.h>
+#include <string.h>
+#include <zephyr/jwt.h>
+#include <entropy.h>
+
+#include <net/tls_credentials.h>
+#include <net/mqtt.h>
+
+#include <mbedtls/platform.h>
+#include <mbedtls/net.h>
+#include <mbedtls/ssl.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/ctr_drbg.h>
+
+#include <mbedtls/debug.h>
+
+#ifdef CONFIG_STDOUT_CONSOLE
+# include <stdio.h>
+# define PRINT printf
+#else
+# define PRINT printk
+#endif
+
+/*
+ * TODO: Properly export these.
+ */
+time_t k_time(time_t *ptr);
+
+/* private key information */
+extern unsigned char zepfull_private_der[];
+extern unsigned int zepfull_private_der_len;
+
+/*
+ * This is the hard-coded root certificate that we accept.
+ */
+#include "globalsign.inc"
+
+static u8_t client_id[] = CONFIG_CLOUD_CLIENT_ID;
+#ifdef CONFIG_CLOUD_SUBSCRIBE_CONFIG
+static u8_t subs_topic_str[] = CONFIG_CLOUD_SUBSCRIBE_CONFIG;
+static struct mqtt_topic subs_topic;
+static struct mqtt_subscription_list subs_list;
+#endif
+static u8_t client_username[] = "none";
+static u8_t pub_topic[] = CONFIG_CLOUD_PUBLISH_TOPIC;
+
+static struct mqtt_publish_param pub_data;
+
+static u8_t token[512];
+
+static bool connected;
+static u64_t next_alive;
+
+/* The mqtt client struct */
+static struct mqtt_client client_ctx;
+
+/* MQTT Broker details. */
+static struct sockaddr_storage broker;
+
+/* Buffers for MQTT client. */
+static u8_t rx_buffer[1024];
+static u8_t tx_buffer[1024];
+
+static sec_tag_t m_sec_tags[] = {
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+		1,
+#endif
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
+		APP_PSK_TAG,
+#endif
+};
+
+void mqtt_evt_handler(struct mqtt_client *const client,
+		      const struct mqtt_evt *evt)
+{
+	struct mqtt_puback_param puback;
+
+	switch (evt->type) {
+	case MQTT_EVT_SUBACK:
+		LOG_INF("[%s:%d] SUBACK packet id: %u", __func__, __LINE__,
+				evt->param.suback.message_id);
+		break;
+
+	case MQTT_EVT_UNSUBACK:
+		LOG_INF("[%s:%d] UNSUBACK packet id: %u", __func__, __LINE__,
+				evt->param.suback.message_id);
+		break;
+
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT connect failed %d", evt->result);
+			break;
+		}
+
+		connected = true;
+		LOG_INF("[%s:%d] MQTT client connected!", __func__, __LINE__);
+
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		LOG_INF("[%s:%d] MQTT client disconnected %d", __func__,
+		       __LINE__, evt->result);
+
+		connected = false;
+
+		break;
+
+#ifdef CONFIG_CLOUD_SUBSCRIBE_CONFIG
+	case MQTT_EVT_PUBLISH:
+		{
+			u8_t d[32];
+			int len = evt->param.publish.message.payload.len;
+
+			mqtt_read_publish_payload(&client_ctx, d, 32);
```

Comment:
There's a slight issue with this part, you should verify how much payload data did you actually recieve. Underlying, it's a non-blocking `recv` call, it might return with no data or not complete data. There's also no guarantee that `sizeof(d)` is larger than payload `len` (though it might be correct in this specific case).

See how it's implemented in one of the MQTT tests:
https://github.com/zephyrproject-rtos/zephyr/blob/d646eff3d37dbcf7cac5821d65f3d1a9dafefcc2/tests/net/lib/mqtt_pubsub/src/test_mqtt_pubsub.c#L118
Note that it's a testcase, hence it has some extra checks. But in general, you should read the payload, until you read `len` bytes of data.

Reply:
> But in general, you should read the payload, until you read len bytes of data.

Yes, that's an absolute requirement, or protocol stream will get unsynchronized. @rlubos, we don't have checks for that, do we?

> Underlying, it's a non-blocking recv call, it might return with no data or not complete data.

Btw, by the end of review of your original PR I got an idea that we need to provide a way to switch MQTT connection into blocking mode, just for the duration of user's reading out the message. I didn't raise it then, to not delay that PR further, but I'm sure we'll need to do that ;-).

Reply:
wait what...... so you're telling me I can receive a publish message and have my event function called, and the mqtt hasn't read out the complete payload?

Reply:
> have my event function called, and the mqtt hasn't read out the complete payload?

Of course not, how mqtt can read the complete payload, if it's 256MB? It's up to user of the API to deal with all those potential megabytes. Streaming support is the key requirement in Zephyr networking APIs.

Reply:
Fun.  It's a great point of failure too.

Reply:
fixed

Reply:
> I can receive a publish message and have my event function called, and we might not have received the full MQTT message

@agross That's how it works now, as @pfalcon says. Main concern here was payload size, we do not limit it now by internal MQTT buffers.

> Yes, that's an absolute requirement, or protocol stream will get unsynchronized. @rlubos, we don't have checks for that, do we?

`mqtt_input` will not allow you to readout a "new" message until you read entire payload of the previous message from a socket, it'll return an error code in that case.

> Btw, by the end of review of your original PR I got an idea that we need to provide a way to switch MQTT connection into blocking mode

Seeing this MQTT implementation used in a few applications now I came up the same conclusion. Having to use `poll` is quite cumbersome in this case.